File: frontend/src/api/compliance.ts
/**
 * 合规初审模块 API 客户端
 */
import api from './client'

// ============ 类型定义 ============

export type ProjectStatus = 'draft' | 'submitted' | 'reviewing' | 'returned' | 'passed' | 'rejected'
export type ReviewResult = 'pass' | 'fail' | 'warning' | 'na'
export type PriorityLevel = 'low' | 'normal' | 'high' | 'urgent'
export type ReviewAction = 'submit' | 'assign' | 'review' | 'return' | 'pass' | 'reject'

export interface ExtractedParam {
  param_name: string
  param_value: string
  source_text?: string
  confidence: string
}

export interface ReviewIssue {
  issue_type: string
  description: string
  severity: string
  source_ref?: string
}

export interface ReviewSuggestion {
  suggestion: string
  related_param?: string
  standard_ref?: string
}

export interface CheckItem {
  id: number
  template_id: number
  category?: string
  item_code?: string
  item_name: string
  item_description?: string
  check_standard?: string
  check_method?: string
  weight: number
  score: number
  is_required: boolean
  risk_level: string
  reference_docs?: any[]
  sort_order: number
  is_active: boolean
  created_at: string
}

export interface ChecklistTemplate {
  id: number
  template_code: string
  template_name: string
  template_type?: string
  template_stage?: string
  description?: string
  version: string
  is_active: boolean
  created_by?: number
  created_at: string
  updated_at: string
  items?: CheckItem[]
  item_count?: number
}

export interface ReviewItem {
  id: number
  checklist_id: number
  check_item_id: number
  category?: string
  item_name: string
  check_standard?: string
  weight: number
  full_score: number
  is_required: boolean
  risk_level: string
  result?: ReviewResult
  score: number
  issue_description?: string
  rectification_suggestion?: string
  evidence?: any[]
  reviewer_note?: string
  reviewed_by?: number
  reviewed_at?: string
  created_at: string
}

export interface ChecklistInstance {
  id: number
  project_id: number
  template_id: number
  template_name: string
  total_items: number
  completed_items: number
  passed_items: number
  failed_items: number
  warning_items: number
  total_score: number
  created_at: string
  review_items?: ReviewItem[]
}

export interface ComplianceReview {
  id: number
  project_id: number
  review_round: number
  action: ReviewAction
  operator_id?: number
  operator_name?: string
  operator_dept?: string
  opinion?: string
  result_data?: any
  attachments?: any[]
  created_at: string
}

export interface Comment {
  id: number
  project_id: number
  parent_id?: number
  comment_type: string
  content: string
  author_id?: number
  author_name?: string
  author_dept?: string
  is_private: boolean
  attachments?: any[]
  created_at: string
  updated_at: string
  replies?: Comment[]
}

export interface Attachment {
  id: number
  project_id: number
  file_name: string
  file_path: string
  file_type?: string
  file_size?: number
  mime_type?: string
  category?: string
  uploader_id?: number
  uploader_name?: string
  description?: string
  is_required: boolean
  created_at: string
}

export interface ComplianceProject {
  id: number
  project_code: string
  project_name: string
  project_type?: string
  project_stage?: string
  applicant?: string
  applicant_dept?: string
  reviewer_id?: number
  reviewer_name?: string
  status: ProjectStatus
  priority: PriorityLevel
  total_score: number
  pass_score: number
  conclusion?: string
  summary?: string
  submitted_at?: string
  reviewed_at?: string
  deadline?: string
  created_by?: number
  created_at: string
  updated_at: string
  checklist_count?: number
  attachment_count?: number
  comment_count?: number
}

export interface ComplianceProjectDetail extends ComplianceProject {
  checklists?: ChecklistInstance[]
  reviews?: ComplianceReview[]
  comments?: Comment[]
  attachments?: Attachment[]
}

export interface ProjectListResponse {
  items: ComplianceProject[]
  total: number
  page: number
  page_size: number
  statistics: {
    all: number
    draft: number
    pending: number
    returned: number
    passed: number
    rejected: number
  }
}

export interface ComplianceStatistics {
  total_projects: number
  draft_count: number
  pending_count: number
  reviewing_count: number
  returned_count: number
  passed_count: number
  rejected_count: number
  avg_score: number
  avg_review_days: number
  pass_rate: number
  by_type: Record<string, number>
  by_month: { month: string; count: number }[]
}

export interface ProjectCreateData {
  project_code: string
  project_name: string
  project_type?: string
  project_stage?: string
  applicant?: string
  applicant_dept?: string
  priority?: PriorityLevel
  pass_score?: number
  deadline?: string
  template_id?: number
}

export interface ProjectUpdateData {
  project_name?: string
  project_type?: string
  project_stage?: string
  applicant?: string
  applicant_dept?: string
  reviewer_id?: number
  reviewer_name?: string
  priority?: PriorityLevel
  pass_score?: number
  deadline?: string
  status?: ProjectStatus
}

export interface ReviewSubmitData {
  action: ReviewAction
  opinion?: string
  reviewer_id?: number
  reviewer_name?: string
  attachments?: any[]
  review_items?: Array<{
    result?: ReviewResult
    score?: number
    issue_description?: string
    rectification_suggestion?: string
    evidence?: any[]
    reviewer_note?: string
  }>
  conclusion?: string
  summary?: string
}

export interface CommentCreateData {
  content: string
  comment_type?: string
  is_private?: boolean
  attachments?: any[]
  parent_id?: number
}

export interface BatchAssignData {
  project_ids: number[]
  reviewer_id: number
  reviewer_name: string
}

// ============ API 函数 ============

// 统计
export const getStatistics = () =>
  api.get<ComplianceStatistics>('/compliance/statistics').then(r => r.data)

// 项目管理
export const createProject = (data: ProjectCreateData) =>
  api.post<ComplianceProject>('/compliance/projects', data).then(r => r.data)

export const getProjects = (params?: {
  page?: number
  page_size?: number
  status?: string
  project_type?: string
  priority?: string
  keyword?: string
  reviewer_id?: number
}) =>
  api.get<ProjectListResponse>('/compliance/projects', { params }).then(r => r.data)

export const getProject = (id: number) =>
  api.get<ComplianceProjectDetail>(`/compliance/projects/${id}`).then(r => r.data)

export const updateProject = (id: number, data: ProjectUpdateData) =>
  api.put<ComplianceProject>(`/compliance/projects/${id}`, data).then(r => r.data)

export const deleteProject = (id: number) =>
  api.delete(`/compliance/projects/${id}`).then(r => r.data)

export const applyChecklistTemplate = (projectId: number, templateId: number) =>
  api.post(`/compliance/projects/${projectId}/submit-checklist`, null, {
    params: { template_id: templateId }
  }).then(r => r.data)

// 审核流程
export const submitReview = (projectId: number, data: ReviewSubmitData) =>
  api.post<ComplianceProjectDetail>(`/compliance/projects/${projectId}/review`, data).then(r => r.data)

export const getReviewHistory = (projectId: number) =>
  api.get<ComplianceReview[]>(`/compliance/projects/${projectId}/reviews`).then(r => r.data)

export const batchAssign = (data: BatchAssignData) =>
  api.post('/compliance/projects/batch-assign', data).then(r => r.data)

// 检查表模板
export const createTemplate = (data: Partial<ChecklistTemplate> & { items?: Partial<CheckItem>[] }) =>
  api.post<ChecklistTemplate>('/compliance/templates', data).then(r => r.data)

export const getTemplates = (params?: {
  is_active?: boolean
  template_type?: string
  keyword?: string
}) =>
  api.get<ChecklistTemplate[]>('/compliance/templates', { params }).then(r => r.data)

export const getTemplate = (id: number) =>
  api.get<ChecklistTemplate>(`/compliance/templates/${id}`).then(r => r.data)

export const updateTemplate = (id: number, data: Partial<ChecklistTemplate>) =>
  api.put<ChecklistTemplate>(`/compliance/templates/${id}`, data).then(r => r.data)

export const deleteTemplate = (id: number) =>
  api.delete(`/compliance/templates/${id}`).then(r => r.data)

// 评论
export const addComment = (projectId: number, data: CommentCreateData) =>
  api.post<Comment>(`/compliance/projects/${projectId}/comments`, data).then(r => r.data)

export const getComments = (projectId: number) =>
  api.get<Comment[]>(`/compliance/projects/${projectId}/comments`).then(r => r.data)

export const updateComment = (commentId: number, data: { content?: string; comment_type?: string; is_private?: boolean }) =>
  api.put<Comment>(`/compliance/comments/${commentId}`, data).then(r => r.data)

export const deleteComment = (commentId: number) =>
  api.delete(`/compliance/comments/${commentId}`).then(r => r.data)

// 附件
export const uploadAttachment = (
  projectId: number,
  file: File,
  options?: {
    file_type?: string
    category?: string
    description?: string
    is_required?: boolean
    onProgress?: (percent: number) => void
  }
) => {
  const formData = new FormData()
  formData.append('file', file)
  if (options?.file_type) formData.append('file_type', options.file_type)
  if (options?.category) formData.append('category', options.category)
  if (options?.description) formData.append('description', options.description)
  if (options?.is_required !== undefined) formData.append('is_required', String(options.is_required))

  return api.post<Attachment>(`/compliance/projects/${projectId}/attachments`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
    onUploadProgress: options?.onProgress
      ? (e) => {
          if (e.total) options.onProgress!(Math.round((e.loaded * 100) / e.total))
        }
      : undefined,
  }).then(r => r.data)
}

export const deleteAttachment = (attachmentId: number) =>
  api.delete(`/compliance/attachments/${attachmentId}`).then(r => r.data)

// 报告
export const generateReport = (projectId: number) =>
  api.get(`/compliance/projects/${projectId}/report`).then(r => r.data)

// ============ 辅助函数 ============

export const getStatusLabel = (status: ProjectStatus): string => {
  const labels: Record<ProjectStatus, string> = {
    draft: '草稿',
    submitted: '待审核',
    reviewing: '审核中',
    returned: '已退回',
    passed: '已通过',
    rejected: '不通过',
  }
  return labels[status]
}

export const getStatusColor = (status: ProjectStatus): string => {
  const colors: Record<ProjectStatus, string> = {
    draft: 'bg-gray-100 text-gray-700',
    submitted: 'bg-blue-100 text-blue-700',
    reviewing: 'bg-yellow-100 text-yellow-700',
    returned: 'bg-orange-100 text-orange-700',
    passed: 'bg-green-100 text-green-700',
    rejected: 'bg-red-100 text-red-700',
  }
  return colors[status]
}

export const getPriorityLabel = (priority: PriorityLevel): string => {
  const labels: Record<PriorityLevel, string> = {
    low: '低',
    normal: '普通',
    high: '高',
    urgent: '紧急',
  }
  return labels[priority]
}

export const getPriorityColor = (priority: PriorityLevel): string => {
  const colors: Record<PriorityLevel, string> = {
    low: 'text-gray-500',
    normal: 'text-blue-600',
    high: 'text-orange-600',
    urgent: 'text-red-600 font-semibold',
  }
  return colors[priority]
}

export const getResultLabel = (result: ReviewResult): string => {
  const labels: Record<ReviewResult, string> = {
    pass: '通过',
    fail: '不通过',
    warning: '警告',
    na: '不适用',
  }
  return labels[result]
}

export const getResultColor = (result: ReviewResult): string => {
  const colors: Record<ReviewResult, string> = {
    pass: 'bg-green-100 text-green-700 border-green-200',
    fail: 'bg-red-100 text-red-700 border-red-200',
    warning: 'bg-yellow-100 text-yellow-700 border-yellow-200',
    na: 'bg-gray-100 text-gray-600 border-gray-200',
  }
  return colors[result]
}

File: backend/app/schemas/compliance.py
"""
合规初审模块 Pydantic Schemas
用于API请求和响应的数据验证
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


# ============ 枚举定义 ============

class ProjectStatus(str, Enum):
    """项目状态枚举"""
    DRAFT = "draft"
    SUBMITTED = "submitted"
    REVIEWING = "reviewing"
    RETURNED = "returned"
    PASSED = "passed"
    REJECTED = "rejected"


class ReviewResult(str, Enum):
    """审核结果枚举"""
    PASS = "pass"
    FAIL = "fail"
    WARNING = "warning"
    NA = "na"


class PriorityLevel(str, Enum):
    """优先级枚举"""
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"
    URGENT = "urgent"


class RiskLevel(str, Enum):
    """风险等级枚举"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class ReviewAction(str, Enum):
    """审核操作枚举"""
    SUBMIT = "submit"
    ASSIGN = "assign"
    REVIEW = "review"
    RETURN = "return"
    PASS = "pass"
    REJECT = "reject"


class CommentType(str, Enum):
    """评论类型枚举"""
    GENERAL = "general"
    PROBLEM = "problem"
    RECTIFICATION = "rectification"
    REPLY = "reply"


# ============ 检查项模板相关 ============

class CheckItemBase(BaseModel):
    category: Optional[str] = Field(None, description="检查类别")
    item_code: Optional[str] = Field(None, description="检查项编码")
    item_name: str = Field(..., description="检查项名称")
    item_description: Optional[str] = Field(None, description="检查项说明")
    check_standard: Optional[str] = Field(None, description="检查标准")
    check_method: Optional[str] = Field(None, description="检查方式")
    weight: float = Field(1.0, description="权重")
    score: float = Field(10.0, description="满分")
    is_required: bool = Field(True, description="是否必查")
    risk_level: str = Field("medium", description="风险等级")
    reference_docs: Optional[List[Dict[str, Any]]] = Field(None, description="参考文档")
    sort_order: int = Field(0, description="排序")


class CheckItemCreate(CheckItemBase):
    pass


class CheckItemUpdate(BaseModel):
    category: Optional[str] = None
    item_code: Optional[str] = None
    item_name: Optional[str] = None
    item_description: Optional[str] = None
    check_standard: Optional[str] = None
    check_method: Optional[str] = None
    weight: Optional[float] = None
    score: Optional[float] = None
    is_required: Optional[bool] = None
    risk_level: Optional[str] = None
    reference_docs: Optional[List[Dict[str, Any]]] = None
    sort_order: Optional[int] = None
    is_active: Optional[bool] = None


class CheckItemResponse(CheckItemBase):
    id: int
    template_id: int
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class ChecklistTemplateBase(BaseModel):
    template_code: str = Field(..., description="模板编码")
    template_name: str = Field(..., description="模板名称")
    template_type: Optional[str] = Field(None, description="适用项目类型")
    template_stage: Optional[str] = Field(None, description="适用项目阶段")
    description: Optional[str] = Field(None, description="模板说明")
    version: str = Field("1.0", description="版本号")


class ChecklistTemplateCreate(ChecklistTemplateBase):
    items: Optional[List[CheckItemCreate]] = Field(None, description="检查项列表")


class ChecklistTemplateUpdate(BaseModel):
    template_code: Optional[str] = None
    template_name: Optional[str] = None
    template_type: Optional[str] = None
    template_stage: Optional[str] = None
    description: Optional[str] = None
    version: Optional[str] = None
    is_active: Optional[bool] = None


class ChecklistTemplateResponse(ChecklistTemplateBase):
    id: int
    is_active: bool
    created_by: Optional[int]
    created_at: datetime
    updated_at: datetime
    items: Optional[List[CheckItemResponse]] = None
    item_count: Optional[int] = None

    class Config:
        from_attributes = True


# ============ 初审项目相关 ============

class ComplianceProjectBase(BaseModel):
    project_code: str = Field(..., description="项目编号")
    project_name: str = Field(..., description="项目名称")
    project_type: Optional[str] = Field(None, description="项目类型")
    project_stage: Optional[str] = Field(None, description="项目阶段")
    applicant: Optional[str] = Field(None, description="申报单位/人")
    applicant_dept: Optional[str] = Field(None, description="申报部门")
    priority: str = Field("normal", description="优先级")
    pass_score: float = Field(60.0, description="及格分数线")
    deadline: Optional[datetime] = Field(None, description="审核截止时间")


class ComplianceProjectCreate(ComplianceProjectBase):
    template_id: Optional[int] = Field(None, description="使用的检查表模板ID")


class ComplianceProjectUpdate(BaseModel):
    project_name: Optional[str] = None
    project_type: Optional[str] = None
    project_stage: Optional[str] = None
    applicant: Optional[str] = None
    applicant_dept: Optional[str] = None
    reviewer_id: Optional[int] = None
    reviewer_name: Optional[str] = None
    priority: Optional[str] = None
    pass_score: Optional[float] = None
    deadline: Optional[datetime] = None
    status: Optional[str] = None


class ComplianceProjectResponse(ComplianceProjectBase):
    id: int
    reviewer_id: Optional[int]
    reviewer_name: Optional[str]
    status: str
    total_score: float
    conclusion: Optional[str]
    summary: Optional[str]
    submitted_at: Optional[datetime]
    reviewed_at: Optional[datetime]
    created_by: Optional[int]
    created_at: datetime
    updated_at: datetime
    checklist_count: Optional[int] = None
    attachment_count: Optional[int] = None
    comment_count: Optional[int] = None

    class Config:
        from_attributes = True


class ComplianceProjectListResponse(BaseModel):
    """项目列表响应（含统计信息）"""
    items: List[ComplianceProjectResponse]
    total: int
    page: int
    page_size: int
    statistics: Optional[Dict[str, Any]] = None


# ============ 检查表实例相关 ============

class ReviewItemBase(BaseModel):
    result: Optional[str] = Field(None, description="审核结果")
    score: float = Field(0.0, description="实际得分")
    issue_description: Optional[str] = Field(None, description="问题描述")
    rectification_suggestion: Optional[str] = Field(None, description="整改建议")
    evidence: Optional[List[Dict[str, Any]]] = Field(None, description="佐证材料")
    reviewer_note: Optional[str] = Field(None, description="审核人备注")


class ReviewItemUpdate(ReviewItemBase):
    pass


class ReviewItemResponse(ReviewItemBase):
    id: int
    checklist_id: int
    check_item_id: int
    category: Optional[str]
    item_name: str
    check_standard: Optional[str]
    weight: float
    full_score: float
    is_required: bool
    risk_level: str
    reviewed_by: Optional[int]
    reviewed_at: Optional[datetime]
    created_at: datetime

    class Config:
        from_attributes = True


class ChecklistInstanceResponse(BaseModel):
    id: int
    project_id: int
    template_id: int
    template_name: str
    total_items: int
    completed_items: int
    passed_items: int
    failed_items: int
    warning_items: int
    total_score: float
    created_at: datetime
    review_items: Optional[List[ReviewItemResponse]] = None

    class Config:
        from_attributes = True


# ============ 审核流程相关 ============

class ComplianceReviewCreate(BaseModel):
    action: str = Field(..., description="操作类型")
    opinion: Optional[str] = Field(None, description="审核意见")
    reviewer_id: Optional[int] = Field(None, description="指派的审核人ID")
    reviewer_name: Optional[str] = Field(None, description="指派的审核人姓名")
    attachments: Optional[List[Dict[str, Any]]] = Field(None, description="附件")
    review_items: Optional[List[ReviewItemUpdate]] = Field(None, description="审核项结果")
    conclusion: Optional[str] = Field(None, description="审核结论")
    summary: Optional[str] = Field(None, description="审核总结")


class ComplianceReviewResponse(BaseModel):
    id: int
    project_id: int
    review_round: int
    action: str
    operator_id: Optional[int]
    operator_name: Optional[str]
    operator_dept: Optional[str]
    opinion: Optional[str]
    result_data: Optional[Dict[str, Any]]
    attachments: Optional[List[Dict[str, Any]]]
    created_at: datetime

    class Config:
        from_attributes = True


# ============ 评论相关 ============

class CommentBase(BaseModel):
    content: str = Field(..., description="评论内容")
    comment_type: str = Field("general", description="评论类型")
    is_private: bool = Field(False, description="是否仅内部可见")
    attachments: Optional[List[Dict[str, Any]]] = Field(None, description="附件列表")


class CommentCreate(CommentBase):
    parent_id: Optional[int] = Field(None, description="回复的父评论ID")


class CommentUpdate(BaseModel):
    content: Optional[str] = None
    comment_type: Optional[str] = None
    is_private: Optional[bool] = None


class CommentResponse(CommentBase):
    id: int
    project_id: int
    parent_id: Optional[int]
    author_id: Optional[int]
    author_name: Optional[str]
    author_dept: Optional[str]
    created_at: datetime
    updated_at: datetime
    replies: Optional[List["CommentResponse"]] = None

    class Config:
        from_attributes = True


# ============ 附件相关 ============

class AttachmentResponse(BaseModel):
    id: int
    project_id: int
    file_name: str
    file_path: str
    file_type: Optional[str]
    file_size: Optional[int]
    mime_type: Optional[str]
    category: Optional[str]
    uploader_id: Optional[int]
    uploader_name: Optional[str]
    description: Optional[str]
    is_required: bool
    created_at: datetime

    class Config:
        from_attributes = True


# ============ 统计相关 ============

class ComplianceStatistics(BaseModel):
    """合规初审统计数据"""
    total_projects: int = Field(0, description="总项目数")
    draft_count: int = Field(0, description="草稿数")
    pending_count: int = Field(0, description="待审核数")
    reviewing_count: int = Field(0, description="审核中数")
    returned_count: int = Field(0, description="退回数")
    passed_count: int = Field(0, description="通过数")
    rejected_count: int = Field(0, description="不通过数")
    avg_score: float = Field(0.0, description="平均得分")
    avg_review_days: float = Field(0.0, description="平均审核天数")
    pass_rate: float = Field(0.0, description="通过率")
    by_type: Dict[str, int] = Field(default_factory=dict, description="按类型统计")
    by_month: List[Dict[str, Any]] = Field(default_factory=list, description="按月份统计")


class ComplianceProjectDetail(ComplianceProjectResponse):
    """项目详情（含关联数据）"""
    checklists: Optional[List[ChecklistInstanceResponse]] = None
    reviews: Optional[List[ComplianceReviewResponse]] = None
    comments: Optional[List[CommentResponse]] = None
    attachments: Optional[List[AttachmentResponse]] = None


# ============ 批量操作 ============

class BatchAssignRequest(BaseModel):
    """批量分配请求"""
    project_ids: List[int]
    reviewer_id: int
    reviewer_name: str


class BatchSubmitRequest(BaseModel):
    """批量提交请求"""
    project_ids: List[int]


# 解决循环引用
CommentResponse.model_rebuild()
